### PR TITLE
Fix memory leaks in Unit Tests

### LINKED ISSUE #5782: Memory leaks found in unit tests
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.0 | Unit tests | Manager/Agent | Packages/Sources | Linux |

## Description

There are some memory leaks in the unit tests found using Valgrind:

### test_analysisd_syscheck 
```
...
==246414== HEAP SUMMARY:
==246414==     in use at exit: 144 bytes in 16 blocks
==246414==   total heap usage: 15,592 allocs, 15,576 frees, 6,786,304 bytes allocated
==246414==
==246414== 144 bytes in 16 blocks are definitely lost in loss record 1 of 1
==246414==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==246414==    by 0x50AB50E: strdup (strdup.c:42)
==246414==    by 0x180B19: setup_fim_data (test_analysisd_syscheck.c:216)
==246414==    by 0x50015F2: ??? (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246414==    by 0x5001E93: _cmocka_run_group_tests (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246414==    by 0x19B829: main (test_analysisd_syscheck.c:3627)
==246414==
==246414== LEAK SUMMARY:
==246414==    definitely lost: 144 bytes in 16 blocks
==246414==    indirectly lost: 0 bytes in 0 blocks
==246414==      possibly lost: 0 bytes in 0 blocks
==246414==    still reachable: 0 bytes in 0 blocks
==246414==         suppressed: 0 bytes in 0 blocks
==246414==
==246414== For lists of detected and suppressed errors, rerun with: -s
==246414== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

### test_enrollment_op
```
...
==246426==
==246426== HEAP SUMMARY:
==246426==     in use at exit: 492 bytes in 17 blocks
==246426==   total heap usage: 6,280 allocs, 6,263 frees, 1,806,669 bytes allocated
==246426==
==246426== 26 bytes in 1 blocks are definitely lost in loss record 1 of 5
==246426==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==246426==    by 0x50AB50E: strdup (strdup.c:42)
==246426==    by 0x1324C4: w_enrollment_cert_init (enrollment_op.c:76)
==246426==    by 0x12B1B6: test_setup_context_3 (test_enrollment_op.c:185)
==246426==    by 0x50015F2: ??? (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246426==    by 0x5001E93: _cmocka_run_group_tests (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246426==    by 0x131E9C: main (test_enrollment_op.c:1061)
==246426==
==246426== 26 bytes in 1 blocks are definitely lost in loss record 2 of 5
==246426==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==246426==    by 0x50AB50E: strdup (strdup.c:42)
==246426==    by 0x1324C4: w_enrollment_cert_init (enrollment_op.c:76)
==246426==    by 0x12B073: test_setup_context_2 (test_enrollment_op.c:165)
==246426==    by 0x50015F2: ??? (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246426==    by 0x5001E93: _cmocka_run_group_tests (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246426==    by 0x131E9C: main (test_enrollment_op.c:1061)
==246426==
==246426== 26 bytes in 1 blocks are definitely lost in loss record 3 of 5
==246426==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==246426==    by 0x50AB50E: strdup (strdup.c:42)
==246426==    by 0x1324C4: w_enrollment_cert_init (enrollment_op.c:76)
==246426==    by 0x12B2E8: test_setup_w_enrolment_request_key (test_enrollment_op.c:204)
==246426==    by 0x50015F2: ??? (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246426==    by 0x5001E93: _cmocka_run_group_tests (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246426==    by 0x131E9C: main (test_enrollment_op.c:1061)
==246426==
==246426== 102 bytes in 2 blocks are definitely lost in loss record 4 of 5
==246426==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==246426==    by 0x50AB50E: strdup (strdup.c:42)
==246426==    by 0x1324AE: w_enrollment_cert_init (enrollment_op.c:75)
==246426==    by 0x12B721: test_setup_enrollment_load_pass (test_enrollment_op.c:249)
==246426==    by 0x50015F2: ??? (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246426==    by 0x5001E93: _cmocka_run_group_tests (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246426==    by 0x131E9C: main (test_enrollment_op.c:1061)
==246426==
==246426== 312 bytes in 12 blocks are definitely lost in loss record 5 of 5
==246426==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==246426==    by 0x50AB50E: strdup (strdup.c:42)
==246426==    by 0x1324C4: w_enrollment_cert_init (enrollment_op.c:76)
==246426==    by 0x12AC95: test_setup_context (test_enrollment_op.c:128)
==246426==    by 0x50015F2: ??? (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246426==    by 0x5001E93: _cmocka_run_group_tests (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246426==    by 0x131E9C: main (test_enrollment_op.c:1061)
==246426==
==246426== LEAK SUMMARY:
==246426==    definitely lost: 492 bytes in 17 blocks
==246426==    indirectly lost: 0 bytes in 0 blocks
==246426==      possibly lost: 0 bytes in 0 blocks
==246426==    still reachable: 0 bytes in 0 blocks
==246426==         suppressed: 0 bytes in 0 blocks
==246426==
==246426== For lists of detected and suppressed errors, rerun with: -s
==246426== ERROR SUMMARY: 5 errors from 5 contexts (suppressed: 0 from 0)
```

### test_run_realtime
```
...
==246474==
==246474== HEAP SUMMARY:
==246474==     in use at exit: 1,784 bytes in 72 blocks
==246474==   total heap usage: 1,916 allocs, 1,845 frees, 215,986 bytes allocated
==246474==
==246474== 12 bytes in 1 blocks are definitely lost in loss record 7 of 41
==246474==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==246474==    by 0x50AB50E: strdup (strdup.c:42)
==246474==    by 0x18B96D: realtime_adddir (run_realtime.c:103)
==246474==    by 0x18730D: test_realtime_adddir_realtime_update (test_run_realtime.c:504)
==246474==    by 0x50014E0: ??? (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246474==    by 0x5001EC4: _cmocka_run_group_tests (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246474==    by 0x189535: main (test_run_realtime.c:1301)
==246474==
==246474== 416 (16 direct, 400 indirect) bytes in 1 blocks are definitely lost in loss record 41 of 41
==246474==    at 0x483DD99: calloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==246474==    by 0x1858C8: setup_group (test_run_realtime.c:64)
==246474==    by 0x50015F2: ??? (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246474==    by 0x5001CDF: _cmocka_run_group_tests (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246474==    by 0x189535: main (test_run_realtime.c:1301)
==246474==
==246474== LEAK SUMMARY:
==246474==    definitely lost: 28 bytes in 2 blocks
==246474==    indirectly lost: 400 bytes in 2 blocks
==246474==      possibly lost: 0 bytes in 0 blocks
==246474==    still reachable: 1,356 bytes in 68 blocks
==246474==         suppressed: 0 bytes in 0 blocks
==246474== Reachable blocks (those to which a pointer was found) are not shown.
==246474== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==246474==
==246474== For lists of detected and suppressed errors, rerun with: -s
==246474== ERROR SUMMARY: 3 errors from 3 contexts (suppressed: 0 from 0)
```

### test_syscheck_audit
```
==246541==
==246541== HEAP SUMMARY:
==246541==     in use at exit: 761,820 bytes in 3,505 blocks
==246541==   total heap usage: 11,042 allocs, 7,537 frees, 2,563,510 bytes allocated
==246541==
==246541== 10 bytes in 1 blocks are definitely lost in loss record 225 of 1,116
==246541==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==246541==    by 0x18B3F3: test_get_process_parent_info_failed (test_syscheck_audit.c:1198)
==246541==    by 0x50014E0: ??? (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246541==    by 0x5001EC4: _cmocka_run_group_tests (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246541==    by 0x194998: main (test_syscheck_audit.c:2607)
==246541==
==246541== 10 bytes in 1 blocks are definitely lost in loss record 226 of 1,116
==246541==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==246541==    by 0x18B5F5: test_get_process_parent_info_passsed (test_syscheck_audit.c:1224)
==246541==    by 0x50014E0: ??? (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246541==    by 0x5001EC4: _cmocka_run_group_tests (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246541==    by 0x194998: main (test_syscheck_audit.c:2607)
==246541==
==246541== LEAK SUMMARY:
==246541==    definitely lost: 20 bytes in 2 blocks
==246541==    indirectly lost: 0 bytes in 0 blocks
==246541==      possibly lost: 0 bytes in 0 blocks
==246541==    still reachable: 761,800 bytes in 3,503 blocks
==246541==         suppressed: 0 bytes in 0 blocks
==246541== Reachable blocks (those to which a pointer was found) are not shown.
==246541== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==246541==
==246541== For lists of detected and suppressed errors, rerun with: -s
==246541== ERROR SUMMARY: 2 errors from 2 contexts (suppressed: 0 from 0)
```

### test_wdb_fim
```
...
==246547==
==246547== HEAP SUMMARY:
==246547==     in use at exit: 6 bytes in 1 blocks
==246547==   total heap usage: 1,055 allocs, 1,054 frees, 118,465 bytes allocated
==246547==
==246547== 6 bytes in 1 blocks are definitely lost in loss record 1 of 1
==246547==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==246547==    by 0x496D319: cJSON_strdup (cJSON.c:170)
==246547==    by 0x49711D2: add_item_to_object (cJSON.c:1926)
==246547==    by 0x49712C9: cJSON_AddItemToObject (cJSON.c:1948)
==246547==    by 0x16E7EB: test_wdb_fim_insert_entry2_fail_element_null (test_wdb_fim.c:241)
==246547==    by 0x50014E0: ??? (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246547==    by 0x5001EC4: _cmocka_run_group_tests (in /usr/lib/x86_64-linux-gnu/libcmocka.so.0.7.0)
==246547==    by 0x1709DE: main (test_wdb_fim.c:466)
==246547==
==246547== LEAK SUMMARY:
==246547==    definitely lost: 6 bytes in 1 blocks
==246547==    indirectly lost: 0 bytes in 0 blocks
==246547==      possibly lost: 0 bytes in 0 blocks
==246547==    still reachable: 0 bytes in 0 blocks
==246547==         suppressed: 0 bytes in 0 blocks
==246547==
==246547== For lists of detected and suppressed errors, rerun with: -s
==246547== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

### PR DESCRIPTION
**Pull request by antoniomanuelfr**
|Related issue|
|---|
|#5782|

## Description
This PR aims to fix several memory errors in the unit tests. In order to improve them, we have added sanitizers when compiling the unit tests, making them fail when a memory leak is detected. This will only work on Linux systems.

Closes #5782 

- Compilation without warnings in every supported platform
  - [X] Linux
  - [X] Windows

### CODE REVIEW

File: src/unit_tests/logcollector/test_logcollector.c [hunk 1]
```diff
@@ -40,140 +40,194 @@ void w_initialize_file_status();
...
-    expect_value(__wrap_OSHash_Add_ex, self, files_status);
-    expect_string(__wrap_OSHash_Add_ex, key, path);
-    will_return(__wrap_OSHash_Add_ex, 0);
-
     int retval = w_update_file_status(path, pos, &context);
-
-    assert_int_equal(retval,-1);
-
+    assert_int_equal(retval, -1);
 }
```
Comment (Molter73) [inline src/unit_tests/logcollector/test_logcollector.c:229]: We should not do any extra changes to tests that need to revisited other than comment them out, please leave the function as it was and comment it out.

File: src/unit_tests/logcollector/test_logcollector.c [hunk 2]
```diff
@@ -863,17 +852,14 @@ void test_w_load_files_status_update_add_fail(void ** state) {
...
-    will_return(__wrap_OSHash_Add_ex, 0);
+    expect_function_call(__wrap_pthread_rwlock_wrlock);
+    expect_function_call(__wrap_pthread_rwlock_unlock);
 
     expect_string(__wrap__merror, formatted_msg, "(1298): Failure to add 'test' to 'file_status' hash table");
 
     w_load_files_status(global_json);
-
 }
+*/
```
Comment (Molter73) [inline src/unit_tests/logcollector/test_logcollector.c:862]: Same as with the previous commented test, don't make any extra changes to it, just comment it out.

File: src/unit_tests/server.cmake [hunk 3]
```diff
@@ -7,8 +7,8 @@ if(NOT WAZUHEXT)
 endif()
 
 # Add compiling flags
-add_compile_options(-ggdb -O0 -g -coverage -DTEST_SERVER -DENABLE_AUDIT -DINOTIFY_ENABLED)
-
+add_compile_options(-ggdb -O0 -g -coverage -DTEST_SERVER -DENABLE_AUDIT -DINOTIFY_ENABLED -fsanitize=address -fsanitize=undefined)
+link_libraries("-fsanitize=address -fsanitize=undefined")
```
Comment (Molter73) [inline src/unit_tests/server.cmake:11]:
```suggestion
link_libraries(-fsanitize=address -fsanitize=undefined)
```

File: src/unit_tests/syscheckd/test_fim_sync.c [hunk 4]
```diff
@@ -412,7 +407,7 @@ static void test_fim_sync_push_msg_no_response(void **state) {
 
 /* fim_sync_checksum */
 static void test_fim_sync_checksum_first_row_error(void **state) {
-    pthread_mutex_t *mutex = NULL;
+    pthread_mutex_t *mutex = (pthread_mutex_t *) 1;
```
Comment (Molter73) [inline src/unit_tests/syscheckd/test_fim_sync.c:410]:
```suggestion
    pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
```

File: src/unit_tests/syscheckd/test_fim_sync.c [hunk 5]
```diff
@@ -422,7 +417,7 @@ static void test_fim_sync_checksum_first_row_error(void **state) {
 }
 
 static void test_fim_sync_checksum_last_row_error(void **state) {
-    pthread_mutex_t *mutex = NULL;
+    pthread_mutex_t *mutex = (pthread_mutex_t *) 1;
```
Comment (Molter73) [inline src/unit_tests/syscheckd/test_fim_sync.c:420]:
```suggestion
    pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
```

File: src/unit_tests/syscheckd/test_fim_sync.c [hunk 6]
```diff
@@ -432,15 +427,15 @@ static void test_fim_sync_checksum_last_row_error(void **state) {
 }
 
 static void test_fim_sync_checksum_checksum_error(void **state) {
-    pthread_mutex_t *mutex = NULL;
+    pthread_mutex_t *mutex = (pthread_mutex_t *) 1;
```
Comment (Molter73) [inline src/unit_tests/syscheckd/test_fim_sync.c:430]:
```suggestion
    pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
```

File: src/unit_tests/syscheckd/test_fim_sync.c [hunk 7]
```diff
@@ -432,15 +427,15 @@ static void test_fim_sync_checksum_last_row_error(void **state) {
...
 
     expect_fim_db_get_data_checksum_error(syscheck.database);
 
     fim_sync_checksum(FIM_TYPE_FILE, mutex);
 }
 
 static void test_fim_sync_checksum_empty_db(void **state) {
-    pthread_mutex_t *mutex = NULL;
+    pthread_mutex_t *mutex = (pthread_mutex_t *) 1;
```
Comment (Molter73) [inline src/unit_tests/syscheckd/test_fim_sync.c:438]:
```suggestion
    pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
```

File: src/unit_tests/syscheckd/test_fim_sync.c [hunk 8]
```diff
@@ -449,13 +444,12 @@ static void test_fim_sync_checksum_empty_db(void **state) {
...
 
-    expect_dbsync_check_msg_call("fim_file", INTEGRITY_CLEAR, 1572521857, NULL, NULL, NULL, strdup("A mock message"));
+    expect_fim_send_sync_control_call("fim_file", INTEGRITY_CLEAR, 1572521857, NULL, NULL, NULL, NULL);
 
-    expect_string(__wrap_fim_send_sync_msg, msg, "A mock message");
     fim_sync_checksum(FIM_TYPE_FILE, mutex);
 }
 static void test_fim_sync_checksum_success(void **state) {
-    pthread_mutex_t *mutex = NULL;
+    pthread_mutex_t *mutex = (pthread_mutex_t *) 1;
```
Comment (Molter73) [inline src/unit_tests/syscheckd/test_fim_sync.c:452]:
```suggestion
    pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
```

**Review by Molter73 — CHANGES_REQUESTED** [review]
Just a few minor changes, other than those the PR is looking good.

File: src/unit_tests/wrappers/wazuh/shared/integrity_op_wrappers.c [hunk 9]
```diff
@@ -32,33 +32,38 @@ char * __wrap_dbsync_state_msg(const char * component, cJSON * data) {
...
+    expect_string(__wrap_fim_send_sync_control, component, component);
+    expect_value(__wrap_fim_send_sync_control, msg, msg);
+    expect_value(__wrap_fim_send_sync_control, id, id);
 
     if (start == NULL) {
-        expect_value(__wrap_dbsync_check_msg, start, 0);
+        expect_value(__wrap_fim_send_sync_control, start, 0);
     } else {
-        expect_string(__wrap_dbsync_check_msg, start, start);
+        expect_string(__wrap_fim_send_sync_control, start, start);
```
Comment (jotacarma90) [inline src/unit_tests/wrappers/wazuh/shared/integrity_op_wrappers.c:51]: This lines needs to be formatted

File: src/unit_tests/logcollector/test_logcollector.c [hunk 10]
```diff
@@ -40,119 +40,173 @@ void w_initialize_file_status();
...
+    free(test_struct->status);
+    free(test_struct);
+
+    return 0;
+}
+
+static int setup_status(void **state) {
+    os_file_status_t *data = NULL;
+
+    if (setup_hashmap(NULL) == 1) {
```
Comment (Molter73) [inline src/unit_tests/logcollector/test_logcollector.c:119]:
```suggestion
    if (setup_hashmap(NULL) != 0) {
```

File: src/unit_tests/logcollector/test_logcollector.c [hunk 11]
```diff
@@ -40,119 +40,173 @@ void w_initialize_file_status();
...
+        return 1;
+    }
+
+    if (data = calloc(1, sizeof(os_file_status_t)), data == NULL) {
```
Comment (Molter73) [inline src/unit_tests/logcollector/test_logcollector.c:123]:
```suggestion
    data = calloc(1, sizeof(os_file_status_t));
    if (data == NULL) {
```

File: src/unit_tests/logcollector/test_logcollector.c [hunk 12]
```diff
@@ -40,119 +40,173 @@ void w_initialize_file_status();
...
+        return 1;
+    }
+
+    test_struct->log_reader->fp = (FILE *) 1;
+    *state = test_struct;
+    return 0;
+}
+
+static int teardown_log_context(void **state) {
+    teardown_hashmap(state);
```
Comment (Molter73) [inline src/unit_tests/logcollector/test_logcollector.c:101]: We should check the return value of this teardown and fail ourselves if it fails. We might also want to move it down so if it fails the rest of the memory is correctly free'd

**Review by Molter73 — CHANGES_REQUESTED** [review]
A few more changes that might have escaped my initial review.

**Review by Molter73 — APPROVED** [review]
LGTM!

**Review by jotacarma90 — DISMISSED** [review]
LGTM

**Review by jotacarma90 — APPROVED** [review]
LGTM

File: src/unit_tests/syscheckd/whodata/test_win_whodata.c [hunk 13]
```diff
@@ -78,7 +78,8 @@ int SIZE_EVENTS;
 
 const PWCHAR WCS_TEST_PATH = L"C:\\Windows\\a\\path";
 const char *STR_TEST_PATH = "c:\\windows\\a\\path";
-
+int __real_OSHash_Add(OSHash *hash, const char *key, void *data);
+void *__real_OSHash_Get(const OSHash *self, const char *key);
```
Comment (Molter73) [inline src/unit_tests/syscheckd/whodata/test_win_whodata.c:82]: Same as a previous comment on `__real_OSHash_Create`, we might want to move this declarations to `hash_op_wrappers.h`

**Review by Molter73 — CHANGES_REQUESTED** [review]
A few comment on where we are placing some declarations, other than those the PR is looking good.

**Review by Molter73 — APPROVED** [review]
LGTM!

**Review by chemamartinez — APPROVED** [review]

**Pull request merged**